File: algorithms/039.Combination_Sum/Combination_Sum.java
public class Solution {
    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> list = new ArrayList<Integer>();
		
        Arrays.sort(candidates);
		combinationSumRecur(candidates,0, target, list, res);
		//System.out.println(res.size());
		//System.out.println(res);
		return res;
    }
	
	public  static void combinationSumRecur(int candidates[],int k,int target,List<Integer> list,List<List<Integer>> res){
		if(target < 0 || k >= candidates.length) return;
		if(0 == target){
			res.add(new ArrayList<>(list));
			return;
		}
		list.add(candidates[k]);
		combinationSumRecur(candidates, k, target-candidates[k], list, res);
		list.remove(list.size()-1);
		while(  k+1 < candidates.length && candidates[k] == candidates[k+1]) k++;
		combinationSumRecur(candidates, k+1,  target, list, res);
	}
}

File: algorithms/039.Combination_Sum/Combination_Sum.py
# Time:  O(k * n^k)
# Space: O(k)

class Solution:
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def combinationSum(self, candidates, target):
        result = []
        self.combinationSumRecu(sorted(candidates), result, 0, [], target)
        return result
    
    def combinationSumRecu(self, candidates, result, start, intermediate, target):
        if target == 0:
            result.append(list(intermediate))
        while start < len(candidates) and candidates[start] <= target:
            intermediate.append(candidates[start])
            self.combinationSumRecu(candidates, result, start, intermediate, target - candidates[start])
            intermediate.pop()
            start += 1

if __name__ == "__main__":
    candidates, target = [2, 3, 6, 7], 7
    result = Solution().combinationSum(candidates, target)     
    print result     
